java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java

```diff
@@ -0,0 +1,255 @@
+          return SymbolicValue.TRUE_LITERAL;
+        } else if (val.equals(SymbolicValue.TRUE_LITERAL)) {
+          return SymbolicValue.FALSE_LITERAL;
+        }
+        // if not tied to a concrete value, create symbolic value with no constraint for now.
+        // TODO : create constraint between expression and created symbolic value
+      default:
+    }
+    return createSymbolicValue(syntaxNode);
+  }
+
+  /**
+   * Remove parenthesis and type cast.
+   */
+  private static Tree skipTrivial(Tree givenExpression) {
+    Tree expression = givenExpression;
+    do {
+      switch (expression.kind()) {
+        case PARENTHESIZED_EXPRESSION:
+          expression = ((ParenthesizedTree) expression).expression();
+          break;
+        case TYPE_CAST:
+          expression = ((TypeCastTree) expression).expression();
+          break;
+        default:
+          return expression;
+      }
+    } while (true);
+  }
+
+  public boolean isNull(ProgramState ps, SymbolicValue val) {
+    return NullConstraint.NULL.equals(ps.constraints.get(val));
+  }
+
+  public boolean isConstrained(ProgramState ps, SymbolicValue val) {
+    return ps.constraints.containsKey(val);
+  }
+
+  public Pair<ProgramState, ProgramState> assumeDual(ProgramState programState, Tree condition) {
+    // FIXME condition value should be evaluated to determine if it is worth exploring this branch. This should probably be done in a
+    // dedicated checker.
+    condition = skipTrivial(condition);
+    switch (condition.kind()) {
+      case INSTANCE_OF: {
+        InstanceOfTree instanceOfTree = (InstanceOfTree) condition;
+        SymbolicValue exprValue = eval(programState, instanceOfTree.expression());
+        if (isNull(programState, exprValue)) {
+          return new Pair<>(programState, null);
+        }
+        // if instanceof is true then we know for sure that expression is not null.
+        return new Pair<>(programState, setConstraint(programState, exprValue, NullConstraint.NOT_NULL));
+      }
+      case EQUAL_TO: {
+        BinaryExpressionTree equalTo = (BinaryExpressionTree) condition;
+        SymbolicValue lhs = eval(programState, equalTo.leftOperand());
+        SymbolicValue rhs = eval(programState, equalTo.rightOperand());
+        if (isNull(programState, lhs)) {
+          ProgramState stateNull = setConstraint(programState, rhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, rhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNotNull, stateNull);
+        } else if (isNull(programState, rhs)) {
+          ProgramState stateNull = setConstraint(programState, lhs, NullConstraint.NULL);
+          ProgramState stateNotNull = setConstraint(programState, lhs, NullConstraint.NOT_NULL);
+          return new Pair<>(stateNotNull, stateNull);
+        }
+        break;
+      }
+      case NOT_EQUAL_TO: {
+        BinaryExpressionTree notEqualTo = (BinaryExpressionTree) condition;
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this line start at column 11. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AIndentationCheck)